File: urbanairship-core/src/main/java/com/urbanairship/iam/InAppActionUtils.java
/* Copyright Airship and Contributors */

package com.urbanairship.iam;

import androidx.annotation.Nullable;

import com.urbanairship.actions.ActionRunRequest;
import com.urbanairship.actions.ActionRunRequestFactory;
import com.urbanairship.json.JsonValue;

import java.util.Map;

/**
 * Action utils for in-app messaging.
 */
public abstract class InAppActionUtils {

    /**
     * Runs actions from the button info.
     *
     * @param buttonInfo The button info.
     */
    public static void runActions(@Nullable ButtonInfo buttonInfo) {
        if (buttonInfo != null) {
            runActions(buttonInfo.getActions());
        }
    }

    /**
     * Runs a map of actions.
     *
     * @param actionsMap The action map.
     */
    public static void runActions(@Nullable Map<String, JsonValue> actionsMap) {
        runActions(actionsMap, null);
    }

    /**
     * Runs a map of actions.
     *
     * @param actionsMap The action map.
     * @param requestFactory Optional action request factory.
     */
    public static void runActions(@Nullable Map<String, JsonValue> actionsMap, @Nullable ActionRunRequestFactory requestFactory) {
        if (actionsMap == null) {
            return;
        }

        for (Map.Entry<String, JsonValue> entry : actionsMap.entrySet()) {
            ActionRunRequest request = requestFactory == null ? ActionRunRequest.createRequest(entry.getKey()) : requestFactory.createActionRequest(entry.getKey());
            request.setValue(entry.getValue()).run();
        }
    }

}

File: urbanairship-automation/src/main/java/com/urbanairship/iam/InAppActionUtils.kt
/* Copyright Airship and Contributors */
package com.urbanairship.iam

import com.urbanairship.actions.Action
import com.urbanairship.actions.ActionRunRequest
import com.urbanairship.actions.ActionRunRequestFactory
import com.urbanairship.iam.info.InAppMessageButtonInfo
import com.urbanairship.json.JsonMap
import com.urbanairship.json.JsonValue

/**
 * Action utils for in-app messaging.
 */
internal object InAppActionUtils {

    /**
     * Runs actions from the button info.
     *
     * @param buttonInfo The button info.
     */
    fun runActions(buttonInfo: InAppMessageButtonInfo?) {
        runActions(buttonInfo?.actions)
    }

    /**
     * Runs a map of actions.
     *
     * @param actionsMap The action map.
     * @param requestFactory Optional action request factory.
     * @param runSync If the actions should be run synchronously.
     */
    @JvmOverloads
    fun runActions(
        actionsMap: JsonMap?,
        requestFactory: ActionRunRequestFactory? = null,
        @Action.Situation situation: Int? = null,
        runSync: Boolean = false
    ) {
        val actions = actionsMap ?: return

        actions.map {
            val result  = requestFactory?.createActionRequest(it.key)
                ?: ActionRunRequest.createRequest(it.key)
            result.apply {
                setValue(it.value)
                situation?.let(this::setSituation)
            }
        }.forEach {
            if (runSync) {
                it.runSync()
            } else {
                it.run()
            }
        }
    }
}
